### PR TITLE
Add MetaShift dataset

### PR DESCRIPTION
**Pull request by dnaveenr**
This PR adds the MetaShift dataset.

Dataset Request  : Add MetaShift dataset [#3813](https://github.com/huggingface/datasets/issues/3813)

@lhoestq As discussed, 
- I have copied the preprocessing script and modified it as required to not create new directories and folders and instead yield the images. 
-  I do the preprocessing in _split_generators to get the required data which is then passed to _generate_examples.
-  Beyond the generated MetaShift dataset, the original preprocess script also generates the meta-graphs for each class, I have currently not included this part. [ Ref : [Link](https://github.com/Weixin-Liang/MetaShift#generate-full-metashift) ]
-  There is a Bonus section, the authors share.  I have currently not included this part. [ Ref : [Link](https://github.com/Weixin-Liang/MetaShift#bonus-generate-the-metashift-attributes-dataset-subsets-defined-by-subject-attributes) ]
- I had a basic test script which downloaded the dataset and tested the basic functionality. Things seems fine.
For real data, I performed the following test : 
```
RUN_SLOW=1 pytest tests/test_dataset_common.py::LocalDatasetTest::test_load_real_dataset_metashift
============================================== test session starts ===============================================
platform linux -- Python 3.7.11, pytest-7.0.1, pluggy-1.0.0
rootdir: ./datasets
plugins: hydra-core-1.1.1, datadir-1.3.1, forked-1.4.0, xdist-2.5.0
collected 1 item                                                                                                 

tests/test_dataset_common.py .                                                                             [100%]

========================================= 1 passed in 4821.25s (1:20:21) =========================================
```
- I couldn't get the dummy dataset. Need some inputs here.
Error as follows : 
```
Using custom data configuration default
Dataset metashift with config None seems to already open files in the method `_split_generators(...)`. You might consider to instead only open files in the method `_generate_examples(...)` instead. If this is not possible the dummy data has to be created with less guidance. Make sure you create the file dummy_data/full-candidate-subsets.pkl.
    for split in generator_splits:
UnboundLocalError: local variable 'generator_splits' referenced before assignment
```

To-Do : 

- [x] Currently I am using the default _SELECTED_CLASSES. I need to use config option here as suggested
- [x] Complete fields in the Dataset Card.
- [x] Tagging the dataset using the Datasets Tagging app. 

Need your help and suggestions for improvement. Thank you

### CODE REVIEW

**Comment by dnaveenr** [comment]
@lhoestq Please could you review this when you get time. Thank you.

**Review by mariosasko — COMMENTED** [review]
Hi! Thanks for working on this! 

> Beyond the generated MetaShift dataset, the original preprocess script also generates the meta-graphs for each class, I have currently not included this part. [ Ref : [Link](https://github.com/Weixin-Liang/MetaShift#generate-full-metashift) ]

Maybe we can add the generated meta-graphs to the card as images (with attributions)?  

> There is a Bonus section, the authors share. I have currently not included this part. [ Ref : [Link](https://github.com/Weixin-Liang/MetaShift#bonus-generate-the-metashift-attributes-dataset-subsets-defined-by-subject-attributes) ]

Would be cool if we could have them as additional configs. Also, maybe we could have configs that expose [image metadata](https://github.com/Weixin-Liang/MetaShift/tree/main/dataset/meta_data) from the `https://nlp.stanford.edu/data/gqa/sceneGraphs.zip` file (this file is downloaded in the script but not used).

> I couldn't get the dummy dataset. Need some inputs here.

I suggest you try to generate the `dataset_infos.json` file first, and then I can help with the dummy data.

**Comment by dnaveenr** [comment]
Thanks a lot for your inputs @mariosasko .
> Maybe we can add the generated meta-graphs to the card as images (with attributions)?

Yes. We can do this for the default set of classes. Will add this.

> Would be cool if we could have them as additional configs. Also, maybe we could have configs that expose [image metadata](https://github.com/Weixin-Liang/MetaShift/tree/main/dataset/meta_data) from the https://nlp.stanford.edu/data/gqa/sceneGraphs.zip file (this file is downloaded in the script but not used).

I'll try adding the bonus section as additional config. 
Regarding exposing the image metadata with a config parameter, how will we showcase/display this information ?

**Comment by dnaveenr** [comment]
Okay. Got it. Will add the these and constants as config parameters.

The image metadata from scene graphs looks like this : 
```json
{
    "2407890": {
        "width": 640,
        "height": 480,
        "location": "living room",
        "weather": none,
        "objects": {
            "271881": {
                "name": "chair",
                "x": 220,
                "y": 310,
                "w": 50,
                "h": 80,
                "attributes": ["brown", "wooden", "small"],
                "relations": {
                    "32452": {
                        "name": "on",
                        "object": "275312"
                    },
                    "32452": {
                        "name": "near",
                        "object": "279472"
                    }                    
                }
            }
        }
    }
}
```
``load_dataset("metashift", selected_classes=["cat", "dog", ...], image_metadata=True)``
How do we showcase/display the image metadata(json) information ?

**Comment by mariosasko** [comment]
> How do we showcase/display the image metadata(json) information ?

We can add the JSON fields as keys to the features dict:
```python
  if self.config.image_metadata:
      features.update({"width": Value("int"), "height": Value("int"), "location": Value("string"), ...}) 
```

P.S. Would rename `image_metadata`  to `with_image_metadata`

**Comment by dnaveenr** [comment]
I have added the following : 
- Added the meta-graphs to the card as images under the Section "Dataset Meta-Graphs".
- Generate the Attributes-Dataset using config parameter. [ [Link](https://github.com/Weixin-Liang/MetaShift#bonus-generate-the-metashift-attributes-dataset-subsets-defined-by-subject-attributes) ]
- Expose image metadata using config parameter.
Format of the image metadata is as follows : [Link](https://cs.stanford.edu/people/dorarad/gqa/download.html)
I have modified the "Objects" which is dict to a list of dicts with as additional parameter named object_id. 
I have defined the structure as follows : 
```
{
    "width": datasets.Value("int64"),
    "height": datasets.Value("int64"),
    "location": datasets.Value("string"),
    "weather": datasets.Value("string"),
    "objects": datasets.Sequence(
        {
            "object_id": datasets.Value("string"),
            "name": datasets.Value("string"),
            "x": datasets.Value("int64"),
            "y": datasets.Value("int64"),
            "w": datasets.Value("int64"),
            "h": datasets.Value("int64"),
            "attributes": datasets.Sequence(datasets.Value("string")),
            "relations": datasets.Sequence(
                {
                    "name": datasets.Value("string"),
                    "object": datasets.Value("string"),
                }
            ),
        }
    ),
}
```
Problem is that objects is not being shown as list of dicts. The output looks as follows : 

> metashift_dataset['train'][0]

```json 
{'image_id': '2338755', 'image': <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=500x281 at 0x7F066C5A49D0>, 'label': 0, 'context': 'ground', 'width': 500, 'height': 281, 'location': None, 'weather': None, 'objects': {'object_id': ['3070704', '3070705', '3070706', '2416713', '3070702', '2790660', '3063157', '2354960', '2037127', '2392939', '2912743', '2125407', '2735257', '3260906', '2351018', '3288269', '3699852', '2734378', '3421201', '2863115'], 'name': ['bicycle', 'bicycle', 'bicycle', 'boot', 'bicycle', 'motorcycle', 'pepperoni', 'head', 'building', 'wall', 'shorts', 'people', 'wheel', 'bricks', 'man', 'cat', 'boot', 'door', 'ground', 'building'], 'x': [137, 371, 458, 215, 468, 399, 368, 245, 0, 140, 260, 284, 138, 451, 339, 187, 210, 26, 0, 313], 'y': [116, 86, 94, 150, 91, 80, 107, 22, 0, 44, 109, 69, 145, 226, 69, 22, 230, 0, 119, 0], 'w': [197, 27, 15, 73, 24, 53, 9, 37, 289, 46, 43, 30, 74, 28, 35, 116, 53, 107, 500, 55], 'h': [126, 25, 38, 128, 43, 50, 16, 44, 158, 73, 51, 52, 97, 15, 73, 252, 46, 147, 162, 77], 'attributes': [[], [], [], ['white'], [], [], [], [], [], [], [], [], [], [], [], ['white'], ['white'], ['large', 'black'], ['brick'], []], 'relations': [{'name': ['to the left of'], 'object': ['3260906']}, {'name': ['to the left of', 'to the right of', 'to the right of', 'to the left of', 'to the right of', 'to the left of', 'to the right of'], 'object': ['3070706', '2351018', '2125407', '2790660', '2037127', '3070702', '3288269']}, {'name': ['to the right of', 'to the right of', 'to the left of', 'to the right of', 'to the right of'], 'object': ['2351018', '3070705', '3070702', '2790660', '3063157']}, {'name': ['to the right of'], 'object': ['2735257']}, {'name': ['to the right of', 'to the right of', 'to the right of', 'to the right of', 'to the right of'], 'object': ['2351018', '2790660', '3070706', '3070705', '3063157']}, {'name': ['to the right of', 'to the right of', 'to the left of', 'to the left of', 'to the right of', 'to the right of', 'to the right of', 'to the right of'], 'object': ['3070705', '2351018', '3070702', '3070706', '3063157', '2125407', '2037127', '3288269']}, {'name': ['to the right of', 'to the left of', 'to the left of', 'to the right of', 'to the right of', 'to the left of', 'to the right of'], 'object': ['2037127', '3070706', '3070702', '2912743', '3288269', '2790660', '2125407']}, {'name': ['to the left of', 'to the right of'], 'object': ['2863115', '2734378']}, {'name': ['to the left of', 'to the left of', 'to the left of', 'to the left of', 'to the left of', 'to the left of'], 'object': ['3070705', '2351018', '3063157', '2125407', '2790660', '2863115']}, {'name': ['to the left of', 'to the right of', 'to the left of'], 'object': ['2125407', '2734378', '3288269']}, {'name': ['to the left of', 'on', 'to the left of'], 'object': ['2351018', '3288269', '3063157']}, {'name': ['to the left of', 'to the left of', 'to the right of', 'to the left of', 'to the right of', 'to the left of'], 'object': ['3063157', '2351018', '2037127', '3070705', '2392939', '2790660']}, {'name': ['to the left of', 'to the left of'], 'object': ['2416713', '3288269']}, {'name': ['to the right of'], 'object': ['3070704']}, {'name': ['to the right of', 'to the left of', 'to the right of', 'to the left of', 'to the left of', 'to the right of', 'to the left of', 'to the right of', 'walking down'], 'object': ['2037127', '2790660', '2125407', '3070705', '3070706', '2912743', '3070702', '3288269', '3421201']}, {'name': ['to the right of', 'to the right of', 'to the left of', 'to the right of', 'to the left of', 'to the left of', 'to the left of', 'to the left of'], 'object': ['2392939', '2734378', '2790660', '2735257', '3063157', '3070705', '2351018', '2863115']}, {'name': [], 'object': []}, {'name': ['of', 'to the left of', 'to the left of', 'to the left of'], 'object': ['2037127', '2354960', '3288269', '2392939']}, {'name': [], 'object': []}, {'name': ['to the right of', 'to the right of', 'to the right of'], 'object': ['2037127', '3288269', '2354960']}]}}
```
Expected output of image_metadata would be : 
```
{'height': 281,
 'location': None,
 'objects': [{'attributes': [],
              'h': 126,
              'name': 'bicycle',
              'object_id': '3070704',
              'relations': [{'name': 'to the left of', 'object': '3260906'}],
              'w': 197,
              'x': 137,
              'y': 116},
             {'attributes': [],
              'h': 25,
              'name': 'bicycle',
              'object_id': '3070705',
              'relations': [{'name': 'to the left of', 'object': '3070706'},
                            {'name': 'to the right of', 'object': '2351018'},
                            {'name': 'to the right of', 'object': '2125407'},
                            {'name': 'to the left of', 'object': '2790660'},
                            {'name': 'to the right of', 'object': '2037127'},
                            {'name': 'to the left of', 'object': '3070702'},
                            {'name': 'to the right of', 'object': '3288269'}],
              'w': 27,
              'x': 371,
              'y': 86},
             {'attributes': ['white'],
              'h': 252,
              'name': 'cat',
              'object_id': '3288269',
              'relations': [{'name': 'to the right of', 'object': '2392939'},
                            {'name': 'to the right of', 'object': '2734378'},
                            {'name': 'to the left of', 'object': '2790660'},
                            {'name': 'to the right of', 'object': '2735257'},
                            {'name': 'to the left of', 'object': '3063157'},
                            {'name': 'to the left of', 'object': '3070705'},
                            {'name': 'to the left of', 'object': '2351018'},
                            {'name': 'to the left of', 'object': '2863115'}],
              'w': 116,
              'x': 187,
              'y': 22},
             {'attributes': ['white'],
              'h': 46,
              'name': 'boot',
              'object_id': '3699852',
              'relations': [],
              'w': 53,
              'x': 210,
              'y': 230},
             .
             .
             .
             {'attributes': ['large', 'black'],
              'h': 147,
              'name': 'door',
              'object_id': '2734378',
              'relations': [{'name': 'of', 'object': '2037127'},
                            {'name': 'to the left of', 'object': '2354960'},
                            {'name': 'to the left of', 'object': '3288269'},
                            {'name': 'to the left of', 'object': '2392939'}],
              'w': 107,
              'x': 26,
              'y': 0},
             {'attributes': ['brick'],
              'h': 162,
              'name': 'ground',
              'object_id': '3421201',
              'relations': [],
              'w': 500,
              'x': 0,
              'y': 119},
             {'attributes': [],
              'h': 77,
              'name': 'building',
              'object_id': '2863115',
              'relations': [{'name': 'to the right of', 'object': '2037127'},
                            {'name': 'to the right of', 'object': '3288269'},
                            {'name': 'to the right of', 'object': '2354960'}],
              'w': 55,
              'x': 313,
              'y': 0}],
 'weather': None,
 'width': 500}

```

May I know how to get the list of dicts representation correctly ?

---
To-Do : 

- [ ] Generate dataset_infos.json file.
- [ ] Add “Dataset Usage” section in the cards and write about the config parameters. 
- [ ] Add the constants that can be tuned as config params.

**Comment by mariosasko** [comment]
> Problem is that objects is not being shown as list of dicts. The output looks as follows :

That's expected. We convert a sequence of dictionaries to a dictionary of sequences to keep the formatting aligned with Tensorflow Datasets. You could disable this behavior by replacing `"objects": datasets.Sequence(object_fields_dict)` with `"objects": [object_fields_dict]`, but that's not what we usually do, so let's keep it like that. 

Also, to limit the size of the dataset repo, please remove the pushed images and pass URLs to the images instead under the `src` attribute (and specify `alt` in case the URLs go down).

I'll do a proper review again after you are finished with the dummy data.

**Comment by dnaveenr** [comment]
> That's expected.
Okay. Got it. Thanks. I thought I was doing something wrong.

> Also, to limit the size of the dataset repo, please remove the pushed images and pass URLs to the images instead under the src attribute (and specify alt in case the URLs go down).
Sure. Where do we host these images ? Can I upload to them to any free image hosting platform or is there any particular website you use ?

**Comment by dnaveenr** [comment]
Update : 
- I have generated the dataset_infos.json file.

> I suggest you try to generate the dataset_infos.json file first, and then I can help with the dummy data.

I am having issues creating the dummy data. I get the following which I use the command : 

`datasets-cli dummy_data datasets/metashift`

```
Dataset metashift with config MetashiftConfig(name='metashift', version=1.0.0, data_dir=None, data_files=None, description=None) seems to already open files in the method `_split_generators(...)`. You might consider to instead only open files in the method `_generate_examples(...)` instead. If this is not possible the dummy data has to be created with less guidance. Make sure you create the file dummy_data/full-candidate-subsets.pkl.
Traceback (most recent call last):
  File "datasets-cli", line 33, in <module>
    sys.exit(load_entry_point('datasets', 'console_scripts', 'datasets-cli')())
  File "/datasets/commands/datasets_cli.py", line 33, in main
    service.run()
  File "/datasets/commands/dummy_data.py", line 324, in run
    dataset_builder=dataset_builder, mock_dl_manager=mock_dl_manager
  File "/datasets/commands/dummy_data.py", line 407, in _print_dummy_data_instructions
    for split in generator_splits:
UnboundLocalError: local variable 'generator_splits' referenced before assignment
```

**Review by lhoestq — COMMENTED** [review]
Feel free to host the images online (on imgur for example) :)

You might have to create the dummy data zip file manually. The `datasets-cli dummy_data ./datasets/metashift` command can help to know which files need to be included in the zip file. In general you just need to include dummy versions of all the files that are downloaded, using the same file names. Dummy zip files must be uncompressed and the directory must be named after the ZIP file name (with the .zip extension)

Let us know if you manage to create them, otherwise we can help

**Comment by dnaveenr** [comment]
> Feel free to host the images online (on imgur for example) :)

Sure. Will do that.

Thanks for the explanation regarding the dummy data zip files. I will try it out and let you know.

**Comment by mariosasko** [comment]
Instead of uploading the images to a hosting service, you can directly reference their GitHub URLs (open the image in the MetaShift repo -> click Download -> copy the image URL). For instance, this is the URL of one of the images:`https://raw.githubusercontent.com/Weixin-Liang/MetaShift/main/docs/figures/Cat-MetaGraph.jpg`. Also, feel free to replace `main` with the most recent commit hash in the copied URLs to make them more robust.

**Comment by dnaveenr** [comment]
@mariosasko I've actually created metagraphs for all the default classes other than those present in the GitHub Repo and included them. :)  The Repo has them only for two classes.

File: datasets/metashift/README.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+languages:
+- en
+licenses:
+- cc-by-4.0
+multilinguality:
+- monolingual
+pretty_name: MetaShift
+size_categories:
+- unknown
+source_datasets: []
```
Comment (lhoestq) [inline datasets/metashift/README.md:15]:
```suggestion
source_datasets:
- original
```

File: datasets/metashift/README.md [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+task_categories:
+- other
+task_ids:
+- other-other-domain-generalization;subpopulation-shifts-settings;assess-training-conflicts
```
Comment (lhoestq) [inline datasets/metashift/README.md:19]:
```suggestion
task_categories:
- image-classification
- other
task_ids:
- multi-label-image-classification
- other-other-domain-generalization
```

File: datasets/metashift/metashift.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+            # Iterate all the subsets of the given subject
+            ##################################
+            for node_name in subject_data:
+                subjects_to_all_set[node_name].update(node_name_to_img_id[node_name])
+
+        return subjects_to_all_set
+
+    @staticmethod
+    def _load_scene_graph(json_path):
+        with open(json_path, "r") as f:
```
Comment (lhoestq) [inline datasets/metashift/metashift.py:304]:
```suggestion
        with open(json_path, "r", encoding="utf-8") as f:
```

**Review by lhoestq — COMMENTED** [review]
Thanks, here are a few suggestions to fix the CI :)

File: datasets/metashift/README.md [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+Consider the following examples to get an idea of how you can use the configuration parameters :
+
+1. To generate the MetaShift Dataset :
+```python
+load_dataset("metashift", selected_classes=['cat', 'dog', 'bus'])
+```
+The full object vocabulary and its hierarchy can be seen [here](https://github.com/Weixin-Liang/MetaShift/blob/main/dataset/meta_data/class_hierarchy.json).
+
+The default classes are _CLASSES = [ 'cat', 'dog', 'bus', 'truck', 'elephant', 'horse',]
```
Comment (mariosasko) [inline datasets/metashift/README.md:88]:
```suggestion
The default classes are `['cat', 'dog', 'bus', 'truck', 'elephant', 'horse']`
```

File: datasets/metashift/README.md [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+2. To generate the MetaShift-Attributes Dataset (subsets defined by subject attributes) :
+```python
+load_dataset("metashift", attributes_dataset = True, attributes=["dog(smiling)", "cat(resting)"])
+```
+
+The default attributes are _ATTRIBUTES = ["cat(orange)", "cat(white)","dog(sitting)","dog(jumping)"]
```
Comment (mariosasko) [inline datasets/metashift/README.md:95]:
```suggestion
The default attributes are `["cat(orange)", "cat(white)", "dog(sitting)", "dog(jumping)"]`
```

File: datasets/metashift/README.md [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+The main advantage is the dataset contains many more coherent sets of data compared to other benchmarks.
+
+Two important benefits of MetaShift :
+- Contains orders of magnitude more natural data shifts than previously available.
+- Provides explicit explanations of what is unique about each of its data sets and a distance score that measures the amount of distribution shift between any two of its data sets.
+
+### Dataset Usage
+
+The dataset has the following configuration parameters:
+- selected_classes: `list[string]`, list of the classes needed to generate the MetaShift dataset.
```
Comment (mariosasko) [inline datasets/metashift/README.md:73]:
```suggestion
- selected_classes: `list[string]`, optional, list of the classes to generate the MetaShift dataset for. If `None`, the list is equal to `['cat', 'dog', 'bus', 'truck', 'elephant', 'horse']`.
```

File: datasets/metashift/metashift.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+        Args:
+        selected_classes: `list[string]`, list of the classes needed to generate subsets.
+        attributes_dataset: `bool`, default=False, to generate the MetaShift-Attributes dataset, set as True
+        attributes: `list[string]`, list of attributes classes needed to generate the Attributes dataset.
+        with_image_metadata: `bool`, default=False, to include image metadata, set as True
+        image_subset_size_threshold: `int`, default=25, The number of images required to consider a subset.
+        If the number of images are less than image_subset_size_threshold, then the subset is not considered.
+        min_local_groups: `int`, default=5, The minimum number of local groups required to consider an object class.
+        **kwargs: keyword arguments forwarded to super.
+        """
```
Comment (mariosasko) [inline datasets/metashift/metashift.py:134]: Please indent the params list here (after `Args:`) and update the docstring with the suggestions from the `Dataset Usage` section to keep these parts as aligned as possible.

**Review by mariosasko — COMMENTED** [review]
I've left a few comments, but other than that looks great.

File: datasets/metashift/metashift.py (hunk 7 above)
Comment (dnaveenr) [reply datasets/metashift/metashift.py]: I've indented the params list and updated the docstring. I haven't included the links mentioned in the Dataset Usage section though, let me know if we need to add them as well.

**Comment by dnaveenr** [comment]
Thanks a lot for your suggestions, Mario. The thing I learnt from the review is that I need to make better sentence formations. I will keep this in mind. :)

**Review by mariosasko — APPROVED** [review]
Not an easy dataset to add, but you did a great job! And it can even be streamed!

Pinging @lhoestq for the final review

File: datasets/metashift/metashift.py [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+            ##################################
+            # TMP: inspect certain class
+            ##################################
+            if subject_str not in subject_classes:
+                continue
+
+            subject_group_summary_dict[subject_str][node_name] = imageID_set_len
+
+        ##################################
+        # Print the subject dict stats
```
Comment (lhoestq) [inline datasets/metashift/metashift.py:274]:
```suggestion
        # Get the subject dict stats
```

File: datasets/metashift/metashift.py [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+_DESCRIPTION = """\
+The MetaShift is a dataset of datasets for evaluating distribution shifts and training conflicts.
+The MetaShift dataset is a collection of 12,868 sets of natural images across 410 classes.
+It was created for understanding the performance of a machine learning model across diverse data distributions.
+"""
+
+_HOMEPAGE = "https://metashift.readthedocs.io/"
+
+
+_LICENSE = "https://github.com/Weixin-Liang/MetaShift/blob/main/LICENSE"
```
Comment (lhoestq) [inline datasets/metashift/metashift.py:71]: here it must be the license of the data

```suggestion
_LICENSE = "Creative Commons Attribution 4.0 International License"
```

**Review by lhoestq — APPROVED** [review]
It looks all good thank you ! I fixed minor issues with the tags and the license

Super impressed by your work on this, congrats :)

**Comment by dnaveenr** [comment]
Thanks a lot for your support. @mariosasko  and @lhoestq .

> Super impressed by your work on this, congrats :)

Its my first dataset contribution to the 🤗 Datasets library, I'm super excited. Thank you. :)

**Pull request merged**